File: ts/project/.sdk/tm/js/src/feature/retry/RetryFeature.js


const { BaseFeature } = require('../base/BaseFeature')


// Automatic retry of transient failures with exponential backoff and
// jitter. Wraps the active transport so a single operation call may make
// several HTTP attempts. A failure is retryable when the transport throws
// / returns an Error, or responds with a status in `statuses`
// (default: 408, 425, 429, 500, 502, 503, 504). An HTTP 429/503 with a
// `Retry-After` header overrides the computed backoff.
class RetryFeature extends BaseFeature {
  version = '0.0.1'
  name = 'retry'
  active = true

  _client
  _options = {}


  init(ctx, options) {
    this._client = ctx.client
    this._options = options || {}
    this.active = options.active

    if (!this.active) {
      return
    }

    const self = this
    const utility = ctx.utility
    const inner = utility.fetcher

    utility.fetcher = async function (ctx2, url, fetchdef) {
      return self._withRetry(ctx2, url, fetchdef, inner)
    }
  }


  async _withRetry(ctx, url, fetchdef, inner) {
    const opts = this._options
    const max = null == opts.retries ? 2 : (opts.retries | 0)
    const minDelay = null == opts.minDelay ? 50 : opts.minDelay
    const maxDelay = null == opts.maxDelay ? 2000 : opts.maxDelay
    const factor = null == opts.factor ? 2 : opts.factor

    let attempt = 0
    let last

    for (; ;) {
      let res
      let threw = false
      try {
        res = await inner(ctx, url, fetchdef)
      }
      catch (err) {
        threw = true
        res = err
      }

      last = res

      const retryable = this._retryable(res)
      if (!retryable || attempt >= max) {
        // Out of attempts: rethrow a thrown error to preserve pipeline
        // semantics, otherwise return the last response/error.
        if (threw && attempt >= max) {
          throw res
        }
        return res
      }

      const wait = this._backoff(res, attempt, minDelay, maxDelay, factor)
      this._track(ctx, attempt + 1, res, wait)
      await this._sleep(wait)
      attempt++
    }
  }


  _retryable(res) {
    if (null == res) {
      return true
    }
    if (res instanceof Error) {
      return true
    }
    const status = res.status
    if ('number' !== typeof status) {
      return false
    }
    const statuses = this._options.statuses || [408, 425, 429, 500, 502, 503, 504]
    return statuses.indexOf(status) >= 0
  }


  _backoff(res, attempt, minDelay, maxDelay, factor) {
    // Honour a server-provided Retry-After (seconds) when present.
    const ra = this._retryAfter(res)
    if (null != ra) {
      return Math.min(maxDelay, ra)
    }
    const base = minDelay * Math.pow(factor, attempt)
    const jitter = false === this._options.jitter ? 0 : Math.floor(Math.random() * minDelay)
    return Math.min(maxDelay, base + jitter)
  }


  _retryAfter(res) {
    if (null == res || null == res.headers) {
      return null
    }
    let v
    if ('function' === typeof res.headers.get) {
      v = res.headers.get('retry-after')
    }
    else {
      v = res.headers['retry-after']
    }
    if (null == v) {
      return null
    }
    const n = Number(v)
    return isNaN(n) ? null : n * 1000
  }


  _sleep(ms) {
    if (null == ms || 0 >= ms) {
      return Promise.resolve()
    }
    const sleep = this._options.sleep
    if ('function' === typeof sleep) {
      return Promise.resolve(sleep(ms))
    }
    return new Promise((r) => setTimeout(r, ms))
  }


  _track(ctx, attempt, res, wait) {
    const client = this._client
    if (null == client._retry) {
      client._retry = { attempts: 0, retries: [] }
    }
    client._retry.attempts++
    client._retry.retries.push({
      attempt,
      status: res && res.status,
      error: res instanceof Error ? res.message : undefined,
      wait,
    })
  }
}


module.exports = {
  RetryFeature
}

File: ts/project/.sdk/tm/swift/Sources/ProjectNameSDK/feature/RetryFeature.swift
// Automatic retry of transient failures with exponential backoff and
// jitter. Wraps the active transport so a single operation call may make
// several HTTP attempts. A failure is retryable when the transport throws,
// returns no value, or responds with a status in `statuses`
// (default: 408, 425, 429, 500, 502, 503, 504). An HTTP 429/503 with a
// `Retry-After` header overrides the computed backoff.

import Foundation

public final class RetryFeature: BaseFeature {
  private var client: ProjectNameSDK?
  private var options: VMap?

  // Activity tracking (mirrors the ts client._retry record).
  public var attempts = 0
  public var retries: [VMap] = []

  public override init() {
    super.init()
    version = "0.0.1"
    name = "retry"
    active = true
  }

  public override func initFeature(_ ctx: Context, _ options: VMap) {
    client = ctx.client
    self.options = options
    active = foptBool(options, "active", false)

    if !active {
      return
    }

    let inner = ctx.utility!.fetcher!

    ctx.utility!.fetcher = { ctx2, url, fetchdef in
      try self.withRetry(ctx2, url, fetchdef, inner)
    }
  }

  private func withRetry(_ ctx: Context, _ url: String, _ fetchdef: VMap,
                         _ inner: FetcherFunc) throws -> Value {
    let max = foptInt(options, "retries", 2)
    let minDelay = foptInt(options, "minDelay", 50)
    let maxDelay = foptInt(options, "maxDelay", 2000)
    let factor = foptNum(options, "factor", 2)

    var attempt = 0

    while true {
      var res: Value = .noval
      var err: Error? = nil
      do {
        res = try inner(ctx, url, fetchdef)
      } catch {
        err = error
      }

      if !retryable(res, err) || attempt >= max {
        // Out of attempts (or not retryable): return the last response as-is
        // (or rethrow) to preserve pipeline semantics.
        if let err = err {
          throw err
        }
        return res
      }

      let wait = backoff(res, attempt, minDelay, maxDelay, factor)
      track(attempt + 1, res, err, wait)
      sleepMs(wait)
      attempt += 1
    }
  }

  private func retryable(_ res: Value, _ err: Error?) -> Bool {
    if err != nil {
      return true
    }
    if isNil(res) {
      return true
    }
    let (status, has) = fresStatus(res)
    if !has {
      return false
    }
    let statuses: [Int]
    if let list = foptList(options, "statuses") {
      statuses = list.items.map { toInt($0) }
    } else {
      statuses = [408, 425, 429, 500, 502, 503, 504]
    }
    for s in statuses {
      if s == status {
        return true
      }
    }
    return false
  }

  private func backoff(_ res: Value, _ attempt: Int, _ minDelay: Int, _ maxDelay: Int,
                       _ factor: Double) -> Int {
    // Honour a server-provided Retry-After (seconds) when present.
    let (ra, has) = retryAfter(res)
    if has {
      return ra > maxDelay ? maxDelay : ra
    }
    let baseWait = Double(minDelay) * pow(factor, Double(attempt))
    var jitter = 0
    if foptBool(options, "jitter", true) && minDelay > 0 {
      jitter = Int.random(in: 0..<minDelay)
    }
    let wait = Int(baseWait) + jitter
    return wait > maxDelay ? maxDelay : wait
  }

  private func retryAfter(_ res: Value) -> (Int, Bool) {
    let (v, has) = fresHeader(res, "retry-after")
    if !has {
      return (0, false)
    }
    let seconds = fparseInt(v, -1)
    if seconds < 0 {
      return (0, false)
    }
    return (seconds * 1000, true)
  }

  private func sleepMs(_ ms: Int) {
    if ms <= 0 {
      return
    }
    foptSleep(options)(ms)
  }

  private func track(_ attempt: Int, _ res: Value, _ err: Error?, _ wait: Int) {
    attempts += 1

    let entry = VMap()
    entry.entries["attempt"] = .int(Int64(attempt))
    entry.entries["wait"] = .int(Int64(wait))
    let (status, has) = fresStatus(res)
    if has {
      entry.entries["status"] = .int(Int64(status))
    }
    if let err = err {
      entry.entries["error"] = .string(errMessage(err))
    }
    retries.append(entry)
  }
}
